review-problems-sol.ml

```diff
@@ -155,13 +156,54 @@ let rec deriv x e = match e with
     | Var(y)-> if y=x then Lit(1.0) else Lit(0.0)
     | Lit(y)-> Lit(0.0)
     | Plus (y,z) -> Plus(deriv x y, deriv x z)
-    | Times (y,z) -> Times(deriv x y, deriv x z)
+    | Times (y,z) -> Plus(Times(y, deriv x z),Times(z, deriv x y))
     | Exp (y) -> Times(Exp(y), deriv x y)
     | Ln (y) -> Times(recip y, deriv x y) 
     
-let nr x0 e = x0-.(eval_at ("x", x0) e)/.(eval_at ("x", x0) (deriv "x" e))
+ let nr x0 e = 
+  let rec f x =
+  let a = (x-.(eval_at ("x", x) e)/.(eval_at ("x", x) (deriv "x" e)))
+    in if (abs_float(a-.0.0))< (0.001) then a else f a 
```

Comment:
`a` is the new estimate for where the root *is*, so it's `f a` that needs to be sufficiently close to zero.
Your logic is otherwise good: if your sufficiently-close check succeeds, you return `a` which is the location of the root. Otherwise you loop around again to further improve the estimate.